Create default OIDC TokenStateManager

sberyozkin opened the pull request:
Fixes #12254

This PR is a new attempt to resolve #12254, following the discussion at the closed #12290.
So, what it does it:
- introduces `SessionManager` interface
- All the existing code dealing with the session cookie is moved to `DefaultSessionManager`
- Added an option for DefaultSessionManager to keep an id token only: if the users don't need access or refresh tokens then the session cookie length is reduced by `/3`. I've been thinking if it can be deduced (for ex we already have a `refresh-expired` property) but came to a conclusion that it will be too messy - for example, even if we check all the injection points for AT and RT, we still won't be able to detect `SecurityIdentity.getCredential(AccessTokenCredential)` etc. So the users can just set the `keep-id-token-only`.

- However, there will always be cases where all 3 tokens are needed, so if a combined cookie value becomes too large then DefaultSessionManager can be configured to create a cookie per token, id-token is still bound to the `q_session` token (the #12290 idea).

- tests are added (custom session manager, DefaultSessionManager keeping the id token only and creating 3 cookies)

I'll follow up with creating an enhancement request to create a DB-based SessionManager - and in total we'll have a very comprehensive support for the OIDC adapter session state management

sberyozkin committed:
Create default OIDC TokenStateManager

sberyozkin commented:
same GraphQL test is failing on Win

sberyozkin commented:
@pedroigor Hey,
By the way, I haven't forgotten about your idea of having a session-cookie, when only id token is needed, keeping the minimum information (ex, name, role, and I'd say also the iat and exp) - but I'd like to see a situation first where just keepin a single id token won't be enough for the default session manager case - in #12254, there is only a small oveflow - but that is related to a cookie containing 3 tokens :-)
But we can certainly follow up on this PR later on, for example, we'd create a new JWT content with a few basic claims only and sign it with the client secret - there will be some extra complexity around, hence I'd prefer to wait till it may be required if possible
thanks

sberyozkin commented:
@gsmet Sorry, did not see you updating the milestones. I'm not sure if this PR will get reviewed/approved fast enough, so I set to 1.10, but it is also a bug fix (without any extra options) so lets see, if can be merged fast enough then I can add a backport label. If the 1.9.0.Final is the right one then please set it again :-)

pedroigor commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java:
```diff
@@ -229,6 +235,45 @@ public void setPostLogoutPath(Optional<String> postLogoutPath) {
+     * Default session manager configuration
+     */
+    @ConfigGroup
+    public static class SessionManager {
```
Maybe rename this to `CookieManager` ?

sberyozkin replied:
@pedroigor I'd prefer not to, please see the comment below

pedroigor commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java:
```diff
@@ -229,6 +235,45 @@ public void setPostLogoutPath(Optional<String> postLogoutPath) {
+         * returned in the authorization code grant response if this property is enabled.
+         */
+        @ConfigItem(defaultValue = "false")
+        public boolean keepIdTokenOnly;
```
Maybe an `enum` such as `CookieStrategy` which `SINGLE`, `SPLIT`, `IDTOKEN`, `ACCESSTOKEN`, `REFRESHTOKEN`.

Where people could choose what type of token should be included in the token. For instance, `ACCESSTOKEN` might be useful if the application wants to relay the token to other services while still tracking session lifetime by either a local introspection or a call to the introspection endpoint.

sberyozkin replied:
@pedroigor I'm really not sure about `SINGLE, SPLIT, IDTOKEN, ACCESSTOKEN, REFRESHTOKEN`. One thing that we always need to keep is the ID token. AT and RT are optional. ID token is what represents the user authentication so we can't ever drop it. 
What would `ACCESSTOKEN` mean - keep ID token and ACCESSTOKEN ? You said the `local introspection` but this is not ID token, they are not really the same thing... We've also had the user issues to do with the binary access tokens not remotely introspectable due to Azure, Google, etc not having an introspection endpoint - you may recall the users started saying their Quarkus apps stopped working after I updated OIDC to verify the code flow AT tokens by default :-)
 
How about this though, I will introduce the enum instead but keep ID token only there - this way we will be able to add more options like ID and access token (combined), etc...
`splitTokens` should still remain a boolean though I think - as it will work with 1 or more tokens combined

pedroigor commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/SessionContent.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Authorization Code Flow Session content
+ */
+public class SessionContent {
```
Perhaps `SessionState` ?

sberyozkin replied:
Hi Pedro @pedroigor So, as said below, the idea is that the session manager returns a `state` which encapsulates somehow the tokens, the default one returns them concatenated by default, but the DB one will return a reference to them etc.

Perhaps instead of `SessionContent` we should simply have `AuthorizationCodeTokens` ? `SessionManager` is translating `AuthorizationCodeTokens` into a session state.

pedroigor commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/SessionManager.java:
```diff
@@ -0,0 +1,15 @@
+ */
+public interface SessionManager {
+
+    String createSessionState(RoutingContext routingContext, OidcTenantConfig oidcConfig, SessionContent sessionContent);
```
Perhaps `SessionManager` should be solely responsible for managing cookies and this method just returns `void`. My point is that we remove any session cookie related code from the mechanism and move here.

I would then just rename these methods to `create`, `get` and `delete`.

sberyozkin replied:
@pedroigor As I've tried to justify, it is actually a `SessionManager` :-)

pedroigor reviewed (commented):
Only minor comments, hope they make sense.

sberyozkin commented:
Hey Pedro @pedroigor Let me do a few changes for us to see how it will look afterwards, but yeah, `SessionManager` or `SessionStateCreator` or similar is what I'm hoping to retain :-)

sberyozkin commented:
Hi @pedroigor I've been thinking about our discussion in this PR while I was offline, so I thought I'd clarify a few more things about how this PR was evolving.
So initially, I started coding `DefaultSessionManager` with the assumption that it would be dealing with the main `q_session` session cookie, but then I realized that it would have to make sure 1) max age, path and domain, secure attributes are all set correctly. Since it is still our code, it may be not too bad. But I don't think we want the custom providers do it too - IMHO `CodeAuthenticationMechanism`  must retain the control - it is already dealing with the code flow redirect state cookie, and post logout.
So then I thought, of course, let this provider just dealing with converting all these tokens in whatever state value it wants. So I did it, simply moved the default `single` strategy there and then also added a `keep-id-token-only`, in fact I was going to open a PR at that point, but then spent another half a day on supporting splitting the tokens - and yes - in this case I had to reuse the code for creating the cookies for AT and RT.
But the whole concept was really about letting non-cookie storage for these tokens be supported.
So, I don't mind renaming it somehow but I'd not like to create a cookie-centric interface here :-), I would not like people who decide to store them say in Vault worry about dealing with the cookies - this is a security sensitive part which better be left with `CodeAuthenticationMechanism`. All the custom implementations should worry about is about returning a value for `CodeAuthenticationMechanism` to keep as a cookie.

But I also like your enum idea - indeed, we can keep enhancing the `default` manager which deals with the cookie to keep pairs like ID+AT or ID+RT, or ID_TRIMMED (your name+roles only) etc, I'm not sure I want this PR address all the variations right now, and as such would prefer to focus on resolving the linked issue, so the enum can start from `single` (or say `id-access-refresh` to make it obvious what it is about) and `id` which should be the 2 most used cases.

I'll start looking at it again most likely next week as I'm currently stuck with something else :-)
Thanks

sberyozkin commented:
Hi Pedro @pedroigor, ok, I've done a few updates:
- renamed `SessionManager` into `TokenStateManager` and updated the docs to mention that `TokenStateManager`'s job is to convert the tokens into a token state which will be set as a session cookie by the coed auth mechanism
- renamed `SesssionContent` to `AuthorizationCodeTokens`
- removed `boolean keepIdTokensOnly` in the default config and added an enum strategy, `KEEP_ALL_TOKENS` (it feels more descriptive than `SINGLE`) is the default, `ID_TOKEN_ONLY` is an option, going forward I can imagine `ID_AND_AT_TOKENS`, etc as needed.

(As I said the default manager does create cookies if it splits them - this splitting can be handled directly in the code auth mechanism, but it is just simpler to keep in the default manager code instead - which is effectively an extension of this mechanism, but of course, it is not something we will be recommending doing in the custom implementations :-)). I hope once we have a DB based provider you will like this interface idea more :-)

I reckon it has become clearer now and more flexible too (your enum strategy idea). 

Have a look please when you get a sec

sberyozkin commented:
@pedroigor I renamed the `id_token_only` enum value to just `id_token`

pedroigor commented:
@sberyozkin Yeah, I think the enum will open space for more options in regards to how the state is stored while making config simpler.

I'm not sure about the `AuthorizationCodeTokens` name but I also don't have a better one :) So, I'm fine with whatever you think is best.

Otherwise, LGTM.

sberyozkin commented:
Hi @pedroigor Thanks, maybe `CodeAuthenticationTokens`, as we already have `CodeAuthenticationMechanism` ?

sberyozkin commented:
We've had a brief chat with Pedro, going to keep `AuthorizationCodeTokens` for now :-)
